src/object/srv_ec.c

```diff
@@ -0,0 +1,296 @@
+					this_recx->rx_nr =
+						oca->u.ec.e_len / iod->iod_size;
+					skip_list[i][sl_idx++] =
+						oca->u.ec.e_len;
+					skip_list[i][sl_idx++] = -(recx_size -
+						(cell_start + oca->u.ec.e_len));
+				} else {
+					this_recx->rx_nr = (recx_size -
+						cell_start) / iod->iod_size;
+					skip_list[i][sl_idx++] = recx_size -
+								 cell_start;
+				}
+			}
+			idx++;
+		}
+	}
+out:
+	return rc;
+}
+
+/* Determines if parity exists for the specified stripe
+ */
+static bool
+ec_has_parity(daos_recx_t *recxs, uint64_t stripe, uint32_t pss,
+	      uint32_t iod_size)
+{
+	unsigned int j;
+
+	for (j = 0; recxs[j].rx_idx & PARITY_INDICATOR; j++) {
+		uint64_t p_stripe = (~PARITY_INDICATOR &
+					(recxs[j].rx_idx * iod_size)) / pss;
+
+		if (p_stripe == stripe)
+			return true;
+	}
+	return false;
+}
+
+/* Process IOD array on parity target. Keeps parity extents that are addressed
+ * to this target. Also retains all data extents that do not have parity for
+ * their stripe.
+ */
+int
+ec_parity_target(unsigned int ptgt_idx, unsigned int nr, daos_iod_t *iods,
+		 struct daos_oclass_attr *oca, long **skip_list)
+{
+	unsigned long	ss = oca->u.ec.e_len * oca->u.ec.e_k;
+	uint32_t	pss = oca->u.ec.e_len * oca->u.ec.e_p;
+	unsigned int	i, j, idx;
+	int		rc = 0;
+
+	for (i = 0; i < nr; i++) {
+		daos_iod_t	*iod = &iods[i];
+		unsigned int loop_bound = iod->iod_nr;
+		int		 sl_idx = 0;
+
+		if (iod->iod_type == DAOS_IOD_SINGLE)
+			continue;
+		D_ALLOC_ARRAY(skip_list[i], iod->iod_nr + 1);
+		if (skip_list[i] == NULL)
+			D_GOTO(out, rc = -1);
+		for (idx = 0, j = 0; j < loop_bound; j++) {
+			daos_recx_t	*this_recx = &iod->iod_recxs[idx];
+
+			if (iod->iod_recxs[idx].rx_idx & PARITY_INDICATOR) {
+				uint64_t	p_address = ~PARITY_INDICATOR &
+					this_recx->rx_idx * iod->iod_size;
+				unsigned int	so = p_address % pss;
+				unsigned int	pcell = so / oca->u.ec.e_len;
+
+				if (pcell == ptgt_idx) {
+					skip_list[i][sl_idx++] =
+						oca->u.ec.e_len;
+				} else {
+					ec_del_recx(iod, idx);
+					skip_list[i][sl_idx++] =
+						-(long)oca->u.ec.e_len;
+					continue;
+				}
+			} else {
+				uint64_t stripe = (this_recx->rx_idx *
+						iod->iod_size) / ss;
+
+				if (ec_has_parity(iod->iod_recxs, stripe, pss,
+						  iod->iod_size)) {
+					ec_del_recx(iod, idx);
+					skip_list[i][sl_idx++] =
+						-(this_recx->rx_nr *
+						  iod->iod_size);
+				} else {
+					skip_list[i][sl_idx++] =
+						(this_recx->rx_nr *
+						 iod->iod_size);
+				}
+			}
+			idx++;
+		}
+	}
+out:
+	return rc;
+}
+
+/* Free the memory allocated on the leader for the copy of the IOD array
+ */
+void
+ec_free_iods(daos_iod_t **iods, int nr)
+{
+	int i;
+
+	for (i = 0; i < nr; i++)
+		D_FREE((*iods)[i].iod_recxs);
+	D_FREE(*iods);
+}
+
+/* Make a copy of the iod array submitted in the RPC. Needed on the leader,
+ * since the original IOD contents must be forwarded to the slave targets.
+ */
+int
+ec_copy_iods(daos_iod_t **out_iod, daos_iod_t *in_iod, int nr)
```

Comment:
a trivial suggestion, output parameter is always the last one of the function, just being consistent with rest part of DAOS